feat(auth): Enhance OAuth callback for robust Docker support

chux0519 opened the pull request:
## TLDR

This pull request improves the OAuth 2.0 authentication flow to make it fully compatible with containerized environments like Docker. It introduces two new environment variables, `OAUTH_CALLBACK_PORT` and `OAUTH_CALLBACK_HOST`, to decouple the server's listening address from the `redirect_uri` sent to Google.

## Dive Deeper
### The Problem

When running `gemini auth login` inside a Docker container, the OAuth flow fails. This is due to a fundamental conflict:

1.  **Google's Requirement**: For security, Google's OAuth server requires the `redirect_uri` to use a loopback address (e.g., `http://localhost:PORT`).
2.  **Docker's Networking**: For a service inside a container to be accessible from the host machine via port mapping, it must listen on `0.0.0.0`, not `localhost`.

The previous implementation used the same host for both constructing the `redirect_uri` and binding the local server, making it impossible to satisfy both conditions simultaneously in Docker.

### The Solution

This PR resolves the conflict by introducing two distinct configurations:

-   `OAUTH_CALLBACK_HOST`: Specifies the IP address for the internal HTTP server to **listen on**. This should be set to `0.0.0.0` when running in Docker.
-   `OAUTH_CALLBACK_PORT`: Specifies a fixed port for both listening and for the `redirect_uri`.

The `redirect_uri` is now always constructed with `localhost`, satisfying Google's security policy, while the server correctly binds to `0.0.0.0` to work with Docker's port mapping.

### How to Use in Docker

This change enables a seamless, one-time login workflow using Docker Compose.

**`docker-compose.yml`:**
```yaml
version: "3.9"

services:
  login:
    image: node:lts
    command: >
      sh -c "
        if [ -f /root/.gemini/oauth_creds.json ]; then
          echo '✅ Credentials already exist.';
        else
          echo '-> Starting interactive login...';
          npx @google/gemini-cli@latest --debug
        fi
      "
    working_dir: /root
    volumes:
      - ./.gemini:/root/.gemini
    ports:
      - "8766:8766" # Map host port to container port
    environment:
      - OAUTH_CALLBACK_PORT=8766
      - OAUTH_CALLBACK_HOST=0.0.0.0
```

**Usage:**
1. Run the one-time, interactive login command:
   ```bash
   docker-compose run --rm --service-ports login
   ```
2. Follow the on-screen instructions to authenticate in your browser.
3. The `oauth_creds.json` file will be created and persisted in the local `./.gemini` directory, ready for use by other services.

### Linked Issues

- Resolves #2040

scottdensmore commented:
@chux0519 can you address the comments then we can get your changes merged in.

scidomino commented on packages/core/src/code_assist/oauth2.ts:
```diff
@@ -117,7 +117,30 @@ export async function getOauthClient(
+    childProcess.on('error', (_) => {
+      // Instead of crashing, we gracefully inform the user to open the URL manually.
+      // The application continues to wait for the callback, so the login can still be completed.
+      console.error(
```
Also: This is better than nothing, but it won't display in the UI unless the user hits "ctrl-s".

scottdensmore commented:
@chux0519 can you resolve the conflict. What OSs did you test on?

chux0519 commented:
Sure I’ll fix them later, I tested on native windows
________________________________
发件人: Scott Densmore ***@***.***>
发送时间: Saturday, July 12, 2025 11:40:58 AM
收件人: google-gemini/gemini-cli ***@***.***>
抄送: Yongsheng Xu ***@***.***>; Mention ***@***.***>
主题: Re: [google-gemini/gemini-cli] feat(auth): Enhance OAuth callback for robust Docker support (PR #3532)

[https://avatars.githubusercontent.com/u/125055?s=20&v=4]scottdensmore left a comment (google-gemini/gemini-cli#3532)<https://github.com/google-gemini/gemini-cli/pull/3532#issuecomment-3064612891>

@chux0519<https://github.com/chux0519> can you resolve the conflict. What OSs did you test on?

―
Reply to this email directly, view it on GitHub<https://github.com/google-gemini/gemini-cli/pull/3532#issuecomment-3064612891>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ADM5S2VOEBHG4BQXLUB73W33IB7UVAVCNFSM6AAAAACBBBQNTCVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTANRUGYYTEOBZGE>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

chux0519 commented:
> ```yaml
> @google/gemini-cli@lates
> ```

Just a quick question about your setup for running the service.

Could you let me know what OS you're running, and if you're using **Windows**, is **Docker Compose** running inside **WSL**? Also, is **10.160.93.26** your host machine's IP address?

Regarding the OAuth setup, the callback URL's host needs to be `localhost` or `127.0.0.1` to comply with the OAuth security policy.

So, you'll need to make sure your container's **port 8766 maps to your host's port 8766**. In your `docker-compose.yaml` file, please add the following environment variables:

```yaml
    environment:
      - OAUTH_CALLBACK_PORT=8766
      - OAUTH_CALLBACK_HOST=0.0.0.0
```

Also, the callback port within the Docker container should also be **8766**.

Since the PR hasn't been merged yet, you can test using my branch. Here's an updated `docker-compose.yaml` snippet you can use:

```yaml
version: "3.9"

services:
  login:
    image: node:lts 
    container_name: gemini_login_helper
    working_dir: /root
    command: >
      sh -c "
        if [ -f .gemini/oauth_creds.json ]; then
          echo '✅ Credentials already exist. No login needed.';
        else
          echo '-> No credentials found. Starting interactive login via npx...';
          npx git+https://github.com/chux0519/gemini-cli.git#oauth_config --debug
        fi
      "
    volumes:
      - ./.gemini:/root/.gemini
    ports:
      - "8766:8766"
    environment:
      - OAUTH_CALLBACK_PORT=8766
      - OAUTH_CALLBACK_HOST=0.0.0.0
```

Let me know if you have any questions\!

cyril-xie commented:
> @cyril-xie
> 
> 这只是关于您运行该服务的设置的一个简单的问题。
> 
> 您能告诉我您运行的是什么操作系统吗？如果您使用的是**Windows**，那么**Docker Compose是否在****WSL**中运行？另外，**10.160.93.26**是您的主机 IP 地址吗？
> 
> 关于 OAuth 设置，回调 URL 的主机需要`localhost`或`127.0.0.1`符合 OAuth 安全策略。
> 
> 因此，您需要确保容器的**端口 8766 映射到主机的端口 8766**。`docker-compose.yaml`请在文件中添加以下环境变量：
> 
> ```yaml
>     environment:
>       - OAUTH_CALLBACK_PORT=8766
>       - OAUTH_CALLBACK_HOST=0.0.0.0
> ```
> 
> 此外，Docker 容器内的回调端口也应该是**8766**。
> 
> 由于此 PR 尚未合并，您可以使用我的分支进行测试。以下是`docker-compose.yaml`您可以使用的更新代码片段：
> 
> ```yaml
> version: "3.9"
> 
> services:
>   login:
>     image: node:lts 
>     container_name: gemini_login_helper
>     working_dir: /root
>     command: >
>       sh -c "
>         if [ -f .gemini/oauth_creds.json ]; then
>           echo '✅ Credentials already exist. No login needed.';
>         else
>           echo '-> No credentials found. Starting interactive login via npx...';
>           npx git+https://github.com/chux0519/gemini-cli.git#oauth_config --debug
>         fi
>       "
>     volumes:
>       - ./.gemini:/root/.gemini
>     ports:
>       - "8766:8766"
>     environment:
>       - OAUTH_CALLBACK_PORT=8766
>       - OAUTH_CALLBACK_HOST=0.0.0.0
> ```
> 
> 如果您有任何疑问，请告诉我！

@chux0519  你好，按照你的方法，我本地的docker可以成功了，但是我部署到服务器上，回调还是不通，比如我的10.160.93.26:8766  这样不是localhost的情况，是否还要再改一下oauth2，因为里面其实还是配置的localhost，是否要改为局域网ip才行

chux0519 commented:
@cyril-xie 这样做是不行的，那个 callback 是需要 host 是 localhost 或是 127.0.0.1 这是个常见的安全策略，你的情况最好是使用端口转发，即把你 server 的端口转发到你本地，你可以查查看 `ssh -L`，做到你访问 localhost:8766 到 你的 $SERVER_IP:8766 就可以

scottdensmore commented:
@chux0519 if you run `npm run preflight` you will see that there are some failing tests. Can you fix these up please 🙏

chux0519 commented:
Sure will fix it
________________________________
发件人: Scott Densmore ***@***.***>
发送时间: Thursday, July 17, 2025 5:53:07 AM
收件人: google-gemini/gemini-cli ***@***.***>
抄送: Yongsheng Xu ***@***.***>; Mention ***@***.***>
主题: Re: [google-gemini/gemini-cli] feat(auth): Enhance OAuth callback for robust Docker support (PR #3532)

[https://avatars.githubusercontent.com/u/125055?s=20&v=4]scottdensmore left a comment (google-gemini/gemini-cli#3532)<https://github.com/google-gemini/gemini-cli/pull/3532#issuecomment-3081204923>

@chux0519<https://github.com/chux0519> if you run npm run preflight you will see that there are some failing tests. Can you fix these up please 🙏

—
Reply to this email directly, view it on GitHub<https://github.com/google-gemini/gemini-cli/pull/3532#issuecomment-3081204923>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ADM5S2TBZQGEGHIJCQ6QTL33I3CUHAVCNFSM6AAAAACBBBQNTCVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTAOBRGIYDIOJSGM>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

chux0519 commented:
@scottdensmore  I reviewed the error and also tested the main branch. I suspect the issue may have been introduced during a previous merge conflict. Since my commit only modified `oauth2.ts`, I’ve cherry-picked my changes onto the latest main branch and force-pushed them to my `oauth_config` branch. This should align the tests with the current state of main.

scottdensmore commented:
@chux0519 still seeing build failures on this PR. 
<img width="2640" height="1652" alt="CleanShot 2025-07-17 at 14 16 51@2x" src="https://github.com/user-attachments/assets/63251ab5-c090-4d8f-8a17-1be3949a9d84" />

chux0519 commented:
@scottdensmore  The issue was caused by a signature mismatch in the mocked http server. It's now fixed.

（by the way, the tests wouldn't run on native Windows, so I switched over to WSL to continue my work.

scottdensmore reviewed (approved):
Approved
